### PR TITLE
Deep consensus update

### PR DESCRIPTION
**Pull request by danielmarchan3**
-Max batch processing to work on bigger datasets
-Training mean val_accuracy threshold to early stop the training 
-Sequential processing instead of parallelization in some importan steps to avoid concurrency
-Fix some bugs

### CODE REVIEW

**Review by cossorzano — DISMISSED** [review]

File: xmipp3/protocols/protocol_screen_deepConsensus.py [hunk 1]
```diff
@@ -609,26 +645,29 @@ def preprocessMicsStep(self):
...
-          print('New mics being preprocessed: ', toPreprocessMicFns)
+          print('New mics to be preprocessed: %d' % len(toPreprocessMicFns))
+          if len(toPreprocessMicFns) > self.PREPROCESS_BATCH_MAX:
+              toPreprocessMicFns = toPreprocessMicFns[:self.PREPROCESS_BATCH_MAX]
+
           if self.ignoreCTF.get():
             preproMicsContent="#mics\n"
-            for micName in micsFnameSet:
-              preproMicsContent+= "%s\n"%(micsFnameSet[micName])
+            for micFileName in toPreprocessMicFns: # CAMBIOS
```
Comment (DStrelak) [inline xmipp3/protocols/protocol_screen_deepConsensus.py:657]: `# CAMBIOS` (multiple occurrences)

File: xmipp3/protocols/protocol_screen_deepConsensus.py [hunk 2]
```diff
@@ -655,122 +694,159 @@ def insertCaculateConsensusSteps(self, mode, prerequisites):
...
-    
+
+    def waitFreeInputMics(self):
+      sl = 0
+      while self.USING_INPUT_MICS:
+        time.sleep(1)
+        sl+=1
+        print('Waiting mics: ', sl)
+        if sl > 60:
+            print('WARNING: a thread has been waiting for more than a minute to acess the input coordinates')
```
Comment (DStrelak) [inline xmipp3/protocols/protocol_screen_deepConsensus.py:705]: `access`

File: xmipp3/protocols/protocol_screen_deepConsensus.py [hunk 3]
```diff
@@ -655,122 +694,159 @@ def insertCaculateConsensusSteps(self, mode, prerequisites):
...
+      self.USING_INPUT_MICS = True
+
     def waitFreeInputCoords(self):
+      sl = 0
       while self.USING_INPUT_COORDS:
         time.sleep(1)
+        sl += 1
+        print('Waiting coords: ', sl)
+        if sl > 60:
+          print('WARNING: a thread has been waiting for more than a minute to acess the input micrographs')
```
Comment (DStrelak) [inline xmipp3/protocols/protocol_screen_deepConsensus.py:716]: `input coordinates`?

File: xmipp3/protocols/protocol_screen_deepConsensus.py [hunk 4]
```diff
@@ -655,122 +694,159 @@ def insertCaculateConsensusSteps(self, mode, prerequisites):
...
-      #Only calculate consensus for coordinates that has not been extracted yet
-      if self.checkIfNewMics(mode):
-        extractedSetOfCoordsFns = []
-        for micFn in self.getExtractedMicFns(mode):
-          extractedSetOfCoordsFns.append(pwutils.path.replaceBaseExt(micFn,'pos'))
+        '''Calculates the consensus coordinates from micrographs whose particles haven't been extracted yet in "mode"'''
+        #Only calculate consensus for coordinates that has not been extracted yet
+        trainedParams = self.loadTrainedParams()
+        if trainedParams['trainingPass'] == '' and mode == 'AND':
+            print('Training already finished')
```
Comment (DStrelak) [inline xmipp3/protocols/protocol_screen_deepConsensus.py:725]:
```
if blabla:
    print ...
    return
no_need_for_long_else_here
```

File: xmipp3/protocols/protocol_screen_deepConsensus.py [hunk 5]
```diff
@@ -655,122 +694,159 @@ def insertCaculateConsensusSteps(self, mode, prerequisites):
...
+                self.USING_INPUT_COORDS = False
+
+                inputFileHeader="#pos_i\n"
+                inputFileStr=inputFileHeader
+                for baseName in inputCoordsFnames:
+                    fnames= inputCoordsFnames[baseName]
+                    inputFileStr+=" ".join(fnames)+"\n"
+
+                assert len(inputFileStr)>len(inputFileHeader), "Error, no consensus can be computed as there " \
+                                                               "are mismatch in coordinate sets filenames"
```
Comment (DStrelak) [inline xmipp3/protocols/protocol_screen_deepConsensus.py:744]: `there is a mismatch in the coordinate set filenames`

File: xmipp3/protocols/protocol_screen_deepConsensus.py [hunk 6]
```diff
@@ -655,122 +694,159 @@ def insertCaculateConsensusSteps(self, mode, prerequisites):
...
-          args = (" -i %(toPickMicsDir)s -c %(toPickCoorsDir)s -o %(outputPosDir)s -s %(boxSize)s " +
-                  "-n %(nToPick)s -t %(nThrs)s") % argsDict
-
-          self.runJob('xmipp_pick_noise', args, numberOfMpi=1)
-
-          self.loadCoords(outputPosDir, 'NOISE', micSet=micNoiseSet)
-          self.TO_EXTRACT_MICFNS['NOISE'] = toPickNoiseFns
-          print('Adding to extract {} {} micrographs'.format(len(toPickNoiseFns), 'NOISE'))
+        trainedParams = self.loadTrainedParams()
+        if trainedParams['trainingPass'] != '':
```
Comment (DStrelak) [inline xmipp3/protocols/protocol_screen_deepConsensus.py:762]: see my previous comment. By switching the order of the branches and `return`, you avoud this long `if`

File: xmipp3/protocols/protocol_screen_deepConsensus.py [hunk 7]
```diff
@@ -655,122 +694,159 @@ def insertCaculateConsensusSteps(self, mode, prerequisites):
...
-                                                               setOfInputCoords=self.inputCoordinates[0].get(),
-                                                               sqliteOutName=sqliteName, write=True)
-          self.USING_INPUT_COORDS=False
-          print("Coordinates %s size: %d" % (mode, totalSetOfCoordinates.getSize()))
-          assert totalSetOfCoordinates.getSize() > MIN_NUM_CONSENSUS_COORDS, \
-            ("Error, the consensus (%s) of your input coordinates was too small (%s). " +
-             "It must be > %s. Try a different input..."
-             ) % (mode, str(totalSetOfCoordinates.getSize()), str(MIN_NUM_CONSENSUS_COORDS))
-          self.coordinatesDict[mode] = totalSetOfCoordinates
+            if not micSet == []:
```
Comment (DStrelak) [inline xmipp3/protocols/protocol_screen_deepConsensus.py:821]: as above. Also, I personally don't like doubly-negative conditions (if not empty)

**Review by DStrelak — CHANGES_REQUESTED** [review]
I don't know, it feels like this file has way too many lines of code to 'just do consensus'

**Comment by danielmarchan3** [comment]
> I don't know, it feels like this file has way too many lines of code to 'just do consensus'

I have the same feeling David, however this protocol does make a lot things different than only a particle picking consensus. Anyway, I will continue working on this protocol, the update was done to make it work on streaming and on big datasets. I am aware that there is still a lot to improve but we are using it in different picking strategies with the ESRF so we need this update in devel branch.

**Review by DStrelak — APPROVED** [review]

**Pull request merged**